xbyak/xbyak_error.h

```diff
@@ -0,0 +1,169 @@
+#pragma once
+#ifndef XBYAK_XBYAK_ERROR_H_
+#define XBYAK_XBYAK_ERROR_H_
```

Comment:
(random place)

Do we have an understanding on how fragile this approach is (if it is, of course)? 

For instance, could it be something like:

``` cpp
void unprotect_page(void *page) {
    int res = mprotect(page, READ|WRITE);
    if (res != OK) throw ERROR("cannot change the page flags");
}

void foo() {
     ...
     unprotect_page(page);
     *page = 1; // if exceptions are turned on, this line cannot fail
                // however, with the "status" approach we can get SegFault here
    ...
}
```

Reply:
This is one of the biggest concern for the approach

Reply:
From what I see, the main drawback of this method compared to exceptions is requirement for manual checking. In order to deal with cases like this, I think we could separate logical and system/runtime errors: logical errors write a status, system/runtime errors cause an `abort()` (or, perhaps, a trap), just like compiling with `-fno-exceptions` would work for `bad_exception`.

Reply:
I have included some changes that may make such cases work by doing a quick return in case of an error. The main issue here is testing, and it may take some time to write a proper test suite for a broad test coverage. However I think we could probably fix issues on a case-by-case basis (or, perhaps, someone could spot a fix earlier than that).

Reply:
There is a test suite that has small coverage wrt exceptions, and at least it works as expected (with modifications) when using no exceptions.

Reply:
I have included two changes that should make us safe:
1. there is now a check if code buffer == 0 in `db(int code)` function which includes code into the buffer;
2. another check for the code buffer when jumping to a label in function `void opJmp(T& label, LabelType type, uint8 shortCode, uint8 longCode, uint8 longPref)` .

According to some manual tests the library produces exception "could not create a primitive". I still get some segfaults, but I think this is because some JIT code is executed that has been created within primitive execution codebase.

Reply:
The changes looks reasonable / acceptable to me. I would suggest to add a test for custom allocator that would fail to check that nothing is missing + substituting the mmap for to emulate the problem on its side. Probably that can be done after the RFC is shown to herumi (but at least adding a note that this is in TODO list before actual PR).

Reply:
I have included a small test with a custom allocators (i) that simply returns NULL, and (ii) makes mmap fail. Thanks for the thorough review.